### PR TITLE
Weight drop

### PR DESCRIPTION
**Pull request by rloganiv**
Hi @scottire,

Can you please merge this into your branch so we can pull it into AllenNLP? There were some pretty subtle bugs in the existing implementations which this addresses. Happy to provide more detail if you want.

Best,

@rloganiv

### CODE REVIEW

**Comment by scottire** [comment]
Thanks for doing this @rloganiv. 

Happy to have a bit more of an explanation so that it's immortalized here for any future readers and because I am not completely clear about why some of the changes were needed. 

My understanding of your changes is:
- To avoid double-counting, the original matched parameters are removed from their parent during init
- The matched parameters are also removed from module dynamically after the forward pass
- A new _Match class encapsulates parameters and their parent modules along with their names.

**Comment by rloganiv** [comment]
Happy to provide an explanation:

The major change I made was to remove the dropped out parameters both at initialization and after each forward pass. This is important for two reasons: 
1. As you've mentioned above this prevents double counting the parameters when initializing an `Optimizer` with `DropConnect.parameters()`.
2. In the some versions of PyTorch applying dropout (with `training=True`) to a tensor will cause it to be a non-leaf tensor. Although using these non-leaf tensors as parameters is neccessary for this module, PyTorch expressly forbids doing this in many parts of its codebase (this is why it is neccessary to override the protected `Module._parameters` dictionary instead of using `Module.register_parameter`). By instantiating and removing the non-leaf parameters within the  forward pass, we avoid a lot of cases where a `ValueError` might be raised.

The other change I added was support for identifying weights using regular expressions. The `_Match` objects were added to help keep track of the matched parameters and their respective parent modules.
